File: backend/app/domain/entities/rag_metrics.py
from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class RagRun:
    id: str
    conversation_id: str
    message_id: str
    user_id: str
    schema_id: str
    schema_name: str
    prompt: str
    status: str
    created_at: datetime
    started_at: datetime
    completed_at: datetime
    duration_ms: int
    attempt_count: int
    repair_count: int
    sql_validated: bool
    sql_executed: bool
    generated_sql: str | None
    error: str | None
    row_count: int
    truncated: bool
    used_context: bool
    context_message_count: int
    model: str
    retrieved_table_count: int
    retrieved_tables: list[str] = field(default_factory=list)


@dataclass(frozen=True)
class RagMetricsFilters:
    from_date: datetime | None = None
    to_date: datetime | None = None
    schema_id: str | None = None
    status: str | None = None
    limit: int = 25


@dataclass(frozen=True)
class RagMetricPoint:
    label: str
    total: int
    successful: int = 0
    failed: int = 0
    average_duration_ms: float = 0


@dataclass(frozen=True)
class RagMetricBucket:
    label: str
    count: int


@dataclass(frozen=True)
class RagSchemaMetric:
    schema_id: str
    schema_name: str
    count: int


@dataclass(frozen=True)
class RagMetricsSummary:
    total_runs: int
    successful_runs: int
    failed_runs: int
    success_rate: float
    average_duration_ms: float
    average_repair_count: float
    average_row_count: float
    runs_by_day: list[RagMetricPoint]
    latency_by_day: list[RagMetricPoint]
    runs_by_schema: list[RagSchemaMetric]
    errors_by_type: list[RagMetricBucket]
    row_count_buckets: list[RagMetricBucket]

File: frontend/src/app/domain/models/rag-metrics.ts
export type RagRunStatus = 'success' | 'error';

export interface RagMetricPoint {
  readonly label: string;
  readonly total: number;
  readonly successful: number;
  readonly failed: number;
  readonly averageDurationMs: number;
}

export interface RagMetricBucket {
  readonly label: string;
  readonly count: number;
}

export interface RagSchemaMetric {
  readonly schemaId: string;
  readonly schemaName: string;
  readonly count: number;
}

export interface RagMetricsSummary {
  readonly totalRuns: number;
  readonly successfulRuns: number;
  readonly failedRuns: number;
  readonly successRate: number;
  readonly averageDurationMs: number;
  readonly averageRepairCount: number;
  readonly averageRowCount: number;
  readonly runsByDay: RagMetricPoint[];
  readonly latencyByDay: RagMetricPoint[];
  readonly runsBySchema: RagSchemaMetric[];
  readonly errorsByType: RagMetricBucket[];
  readonly rowCountBuckets: RagMetricBucket[];
}

export interface RagRun {
  readonly id: string;
  readonly createdAt: Date;
  readonly status: RagRunStatus;
  readonly schemaId: string;
  readonly schemaName: string;
  readonly conversationId: string;
  readonly messageId: string;
  readonly prompt: string;
  readonly generatedSql: string | null;
  readonly error: string | null;
  readonly durationMs: number;
  readonly attemptCount: number;
  readonly repairCount: number;
  readonly rowCount: number;
  readonly usedContext: boolean;
}

export interface RagMetricsFilters {
  readonly from?: string;
  readonly to?: string;
  readonly schemaId?: string;
}
